### DISCUSSION TITLE
FR: GraphQL type entries by entry type where possible instead of EntryInterface

### DISCUSSION THREAD

**Discussion by MangoMarcus** [craftcms/cms#14372] [Ideas] [answered]
I have entry type, `blog`, which has an `entries` field, `person`, which is restricted to entries in section, `people`.

The problem is that when I query `person` it's typed as an `EntryInterface` which could be of any entry type, rather than just those in the `people` section.

I'm using [graphql-codegen](https://the-guild.dev/graphql/codegen) in a headless app to generate Typescript types from my GQL queries, which use Craft's inbuilt `entries` query.

Because entries aren't typed to their section, I get type errors when referencing fields which are unique to a specific entry type.

My workaround is to extract the desired type from the union. Eg. if type `Query['entries'][0]['person']` is

```ts
{ __typename: 'people_default_Entry', /* ... people fields */ }
| { __typename: 'blog_default_Entry', /* ... blog fields */ }
| { __typename: 'home_home_Entry', /* ... home fields */ }
// ... other entry types
```

Then the correct type can be defined using
```ts
type Person = Extract<Query['entries'][0]['person'], { __typename: 'people_default_Entry' }>
```

And then use that as a type assertion

```ts
const person = entry.person as Person;
person.email; // No type error!
```

This does the job but adds some maintenance overhead for managing the types so I'm wondering if there's a better way. My knowledge of GQL schema set up but would it be possible to type an entry field like this to only its selected "sources"?

Alternatively let me know if I've missed something or I'm going about it the wrong way, thanks!

**Comment by brandonkelly** [comment]
I’m not familiar with graphql-codegen, but the normal GraphQL way of handling this would be with a fragment, e.g.

```graphql
query {
  blogEntries {
    ...on blog_default_Entry {
      person {
        title
        ...on people_default_Entry {
          // people fields
        }
      }
    }
  }
}
```

**Reply by brandonkelly** [reply]
Entries fields don’t know which entry types they might end up pulling in, so a union type would just end up including every possible entry type, which would be pretty inefficient.

**Reply by MangoMarcus** [reply]
Thanks @brandonkelly ,

> Entries fields don’t know which entry types they might end up pulling in]

Is it not possible to infer this based on the sources selected in the field's settings? Eg. below

Sorry if a noob question, thanks

<img width="445" alt="Screenshot 2024-02-14 at 12 12 39" src="https://github.com/craftcms/cms/assets/72021142/45e7aef2-6171-47de-aa7f-3a7924702b49">

**Reply by brandonkelly** [reply, marked as answer]
That list isn’t limited to sections; it could include custom sources, or plugin-defined sources, so yeah, we just don’t know for sure.

**Reply by MangoMarcus** [reply]
Ahh OK that makes sense, ty for the clarification @brandonkelly !
